internal/scheduler/scheduling_algo.go

```diff
@@ -91,16 +101,45 @@ func (l *FairSchedulingAlgo) Schedule(
 	overallSchedulerResult := &SchedulerResult{
 		NodeIdByJobId: make(map[string]string),
 	}
-	for _, executor := range accounting.executors {
+
+	executorsToSchedule := accounting.executors
+	slices.SortStableFunc(executorsToSchedule, func(a, b *schedulerobjects.Executor) bool {
+		return strings.Compare(a.Id, b.Id) < 1
+	})
+
+	schedCtx, cancel := context.WithTimeout(ctx, l.maxSchedulingDuration)
+	defer cancel()
+
+	allExecutorsConsidered := false
+	for i, executor := range executorsToSchedule {
+		if schedCtx.Err() != nil {
+			// We've reached the scheduling time limit, exit gracefully
+			break
+		}
+
+		if i+1 == len(executorsToSchedule) {
```

Comment:
this doesn't work quite how I imagined it.  I'd assumed we'd always try and schedule at least `len(executorsToSchedule)`. I.e. assume we have 4 executors, I would asssume:
```
Scheduling round 1: schedules on 1,2,3 (times out)
Scheduling round 2: schedules on 4,1,2,3 (finishes after scheduling on n executors)
```
Whereas I think this code would do:
```
Scheduling round 1: schedules on 1,2,3 (times out)
Scheduling round 2: schedules on 4 (finishes after we reach the end of the slice)
```